docs/site/Context.md

```diff
@@ -530,3 +530,101 @@ class MyController {
   }
 }
 ```
+
+## Configuration by convention
+
+To allow bound items in the context to be configured, we introduce some
+conventions and corresponding APIs to make it simple and consistent.
+
+We treat configurations for bound items in the context as dependencies, which
+can be resolved and injected in the same way of other forms of dependencies. For
+example, the `RestServer` can be configured with `RestServerConfig`.
+
+Let's first look at an example:
+
+```ts
+export class RestServer {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) app: Application,
+    @inject(RestBindings.CONFIG, {optional: true})
+    config: RestServerConfig = {},
+  ) {
+    // ...
+  }
+  // ...
+}
+```
+
+The configuration (`RestServerConfig`) itself is a binding
+(`RestBindings.CONFIG`) in the context. It's independent of the binding for
+`RestServer`. The caveat is that we need to maintain a different binding key for
+the configuration. Referencing a hard-coded key for the configuration also makes
+it impossible to have more than one instances of the `RestServer` to be
+configured with different options, such as `protocol` or `port`.
+
+To solve these problems, we introduce an accompanying binding for an item that
+expects configuration. For example:
+
+- `servers.RestServer.server1`: RestServer
+- `servers.RestServer.server1.$config`: RestServerConfig
+
+- `servers.RestServer.server2`: RestServer
+- `servers.RestServer.server2.$config`: RestServerConfig
+
+The following APIs are available to enforce/leverage this convention:
+
+1. ctx.configure('servers.RestServer.server1') => Binding for the configuration
+2. ctx.getConfig('servers.RestServer.server1') => Get configuration
+3. `@config` to inject corresponding configuration
+4. `@config.getter` to inject a getter function for corresponding configuration
+5. `@config.view` to inject a `ContextView` for corresponding configuration
```

Comment:
Please explain that:

All configuration accessors (`ctx.getConfig`, `@config*`) consider the configuration binding as optional, i.e. return `undefined` if no configuration was bound. This is different from `ctx.get` and `@inject` APIs, which treat bindings as required and throw an error when the requested binding is not found.

Reply:
> >  Please explain that:
>
> All configuration accessors (`ctx.getConfig`, `@config*`) consider the configuration binding as optional, i.e. return `undefined` if no configuration was bound. This is different from `ctx.get` and `@inject` APIs, which treat bindings as required and throw an error when the requested binding is not found.

I see that you have documented this binding in API docs. Can you please describe it in `docs/site/Context.md` too? In my experience, most people are not going to read our API docs.